### PR TITLE
[tf.data] Use output_shapes from python for batch dataset

### PR DESCRIPTION
**Pull request by zhuzilin**
This is a PR from JIZHI, the AI platform in Tencent.

This pr is related to #40938 . It removes the output shape calculation in C++ for `BatchDatasetOp` and `PaddedBatchDatasetOp` and use the shapes passed from python instead.

Thank you for your time on reviewing this pr.

### CODE REVIEW

**Comment by zhuzilin** [comment]
@jsimsa 
Could you have a look at this pr? Thank you!

**Comment by zhuzilin** [comment]
@jsimsa
Could you have a look at this pr? Thank you!

**Comment by jsimsa** [comment]
Hi @zhuzilin sorry for the delay in response. I was out of office for an extended period of time and am still catching up.

What is the motivation for your PR? I would prefer to keep the shape inference in C++ for the following reasons.

In general, we could rewrite the tf.data graph in a way that would allow the C++ shape inference to be more accurate. For instance, if the rewrite would set the `drop_remainder` attribute to `True`. If use the shapes from the original Python shape inference, than we lose this ability.

In other words, there are reasons for having separate C++ and Python shape inference and I would prefer to keep both.

**Comment by zhuzilin** [comment]
@jsimsa 
Thank you for your nice explanation! This pr is just me wondering why we need both c++ version and python version of the shape inference.
